features/256-auth-gear-interaction-overview-in-next/api.md

```diff
@@ -172,25 +176,38 @@ Followings are hooks of auth actions:
-| `verify` | `before_verify_sync(user, context)`<br/>`before_verify(user, context)`<br/>`after_verify_sync(user, context)`<br/>`after_verify(user, context)` |
-| `unverify` | `before_unverify_sync(user, context)`<br/>`before_unverify(user, context)`<br/>`after_unverify_sync(user, context)`<br/>`after_unverify(user, context)` |
-| `update_metadata` | `before_update_metadata_sync(user, original_user, context)`<br/>`before_update_metadata(user, original_user, context)`<br/>`after_update_metadata_sync(user, original_user, context)`<br/>`after_update_metadata(user, original_user, context)` | 
+| `roles` | `before_roles_changed_sync(user, original_user, context)`<br/>`before_roles_changed(user, original_user, context)`<br/>`after_roles_changed_sync(user, original_user, context)`<br/>`after_roles_changed(user, original_user, context)` |
+| `enable` | `before_enable_changed_sync(user, original_user, context)`<br/>`before_enable_changed(user, original_user, context)`<br/>`after_enable_changed_sync(user, original_user, context)`<br/>`after_enable_changed(user, original_user, context)` |
+| `password` | `before_password_changed_sync(user, original_user, context)`<br/>`before_password_changed(user, original_user, context)`<br/>`after_password_changed_sync(user, original_user, context)`<br/>`after_password_changed(user, original_user, context)` |
+| `verify` | `before_verify_changed_sync(user, original_user, context)`<br/>`before_verify_changed(user, original_user, context)`<br/>`after_verify_changed_sync(user, original_user, context)`<br/>`after_verify_changed(user, original_user, context)` |
+| `metadata` | `before_metadata_changed_sync(user, original_user, context)`<br/>`before_metadata_changed(user, original_user, context)`<br/>`after_metadata_changed_sync(user, original_user, context)`<br/>`after_metadata_changed(user, original_user, context)` |
+| `user` | `before_user_changed_sync(user, original_user, context)`<br/>`before_user_changed(user, original_user, context)`<br/>`after_user_changed_sync(user, original_user, context)`<br/>`after_user_changed(user, original_user, context)` |
+
+1. Hooks listed presented here are based on a assumption that a developer could use `content.req.path` to know the reason of certain user auth data changed.
+
+   For example, a user's password changed could be due to:
+
+   1. admin reset a user's password.
+   2. a user changes password by himself.
+   3. a user forgot password, and reset password via email form.
+
+   They are corresponding to following `context.req.path`:
+
+   1. `/reset_password`
+   2. `/change_passowrd`
+   3. `/forgot_password/reset_password`
+
+2. `user_changed` hooks are useful when a developer isn't sure which hook to implement, or it can be a hook to centralize a business logic that may be crossed over many hooks.
```

Comment:
I don't think this is a good description. A better description should be:

2. `user_profile_sync` / `user_profile_async` are special hooks for a developer to implement an external user profile store and want ONE hook to capture all changes to the user auth data or metadata.